### ISSUE TITLE
Detect duplicates in Flow Validation

### ISSUE THREAD

**Issue by carmine** [telstra/open-kilda#615]
We've found times when either there are two entries in the flow table (ie same direction, two flows).  The flow validation should detect this and report this error.  Even better, detect, report, and identify which one is the right one (by comparing to FT memory).

Additionally, there have been scenarios where the rule on the switch is out of sync, based on an error or something else.  Scenarios to detect and look for:

- multiple cookies ... right now we expect the cookie to be unique (it should be) .. but if there is an error, there could be multiple, and flow validation only chooses the first or whatever.
- the wrong cookie ... so, whereas the least significant 4 bytes might be the same ... the significant bits could be off. This is a worthwhile check in scenario where we can't find cookie but can find in port / in vlan.

**Comment by pzakatov** [comment]
@sergii-iakovenko , is this still actual?

**Comment by sergii-iakovenko** [comment]
@pzakatov Yes. Neither flow nor switch validation don't check cookies for duplication.

**Comment by Settis** [comment]
Change description with Proposal.

**Comment by pzakatov** [comment]
@carmine @jonvestal @surabujin please review the proposal. Thanks.

**Comment by surabujin** [comment]
To avoid adding storage for intermediate data(dump of OFRules for each switch on flow path) in storm we should do flow validation in following way:
* in NB fetch flow path
* for each switch in flow path, send command `validate flow on switch` (handled by nbworker topology)
* collect all validation results
* assemble and return response

I.e NB will keep intermediate results and contain logic to assemble validation chunks in whole response.

**Comment by carmine** [comment]
Would it make sense to hold any of this in a database table, and only for requested validation?

Doing it all in NB is okay; it keeps it ephemeral; if NB dies, the client would have to request again (this is okay); however, we don't have any record that such a request was made and what the result was. Keeping this in a table allows us to:
1) know what was requested, and what the result was
2) when the last time such a request was made
3) we can extend and keep a record of failures.

Thoughts?

**Comment by carmine** [comment]
Regarding the core logic above, I think it is sound. However, I'd like to see this put into a library so that it can be shared between NB and a Janitor (who would periodically validate flows in the same way). Additionally, this should be part of the final step of the hub and spoke model for flow crud.

**Comment by Settis** [comment]
Description was updated.

**Comment by Settis** [comment]
Description was updated. Information about refactoring was moved into separate issues: #1441 and #1442.

**Comment by nikitamarchenko** [comment]
@rtretyak

**Comment by rtretyak** [comment]
Will do
